### PR TITLE
Auth

### PR DESCRIPTION
**Pull request by adelajaOlamilekan**
​
## Description
This commit improved the authentication to save a user session when logged in and also provides a service to check if a user is logged in.

​
## Related Issue (Link to linear ticket)

​
## Motivation and Context
This change is required to ensure our app complies with the standard way of authentication where users login state is saved to ensure persistence of user session

​
## How Has This Been Tested?
This has been tested by opening the swagger docs on two different browsers with different signed and testing if their session were independent of each other (i.e ending one user session doesn't end every users session). I also tested that a logged in user get a prompt that he or she is logged in if the person tries to re-login. I also ensured a user that is not logged in cannot logout (by letting them they are not logged in)

​
## Screenshots (if appropriate - Postman, etc):
Screenshot (1258)
Screenshot (1259)
Screenshot (1260)
Screenshot (1261)
Screenshot (1262)
​
## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
​
## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

File: app/routes/auth_routes.py [hunk 1]
```diff
@@ -1,23 +1,82 @@
...
+    client_kwargs={'scope': 'openid email profile'},
+)
+
+# Set up the middleware to read the request session
+# SECRET_KEY = os.environ.get('SECRET_KEY') or None
+# if SECRET_KEY is None:
+#     raise 'Missing SECRET_KEY'
+
+# auth_router.add_middleware(SessionMiddleware, secret_key=SECRET_KEY)
+
```
Comment (zxenonx) [inline app/routes/auth_routes.py:60]: Could we remove this?

File: app/services/services.py [hunk 2]
```diff
@@ -212,3 +213,18 @@ def generate_id():
...
+    Parameters:
+    - request: Holds the request metadata of a user when interacting with the app.
+
+    Returns:
+    - A truthy of Falsy value indicating if user is currently logged in or not 
+    """
+    if "username" in request.session and "logged_in" in request.session:
+        return True
+    else:
+        return False
```
Comment (zxenonx) [inline app/services/services.py:230]:
```suggestion
        return "username" in request.session and "logged_in" in request.session
```
the condition is already a boolean expression. Do you get? @adelajaOlamilekan

File: app/services/services.py [hunk 3]
```diff
@@ -212,3 +213,18 @@ def generate_id():
     
     ALPHABET = "0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz"
     return str(nanoid.generate(size=10, alphabet=ALPHABET))
+
+def is_loggedin(request: Request) -> bool:
```
Comment (zxenonx) [inline app/services/services.py:217]: I suggest "is_logged_in"

File: app/routes/auth_routes.py [hunk 4]
```diff
@@ -1,23 +1,82 @@
...
+
+oauth2_scheme = OAuth2PasswordBearer(tokenUrl="token")
+
+
+@auth_router.get('/')
+def public(request: Request):
```
Comment (zxenonx) [inline app/routes/auth_routes.py:66]: Why is it named "public"?

File: app/routes/auth_routes.py [hunk 5]
```diff
@@ -49,16 +108,21 @@ async def signup_user(user: UserRequest, db: Session= Depends(get_db)):
...
+    
+    #checking if the user is currently logged in
+    user_is_loggedin = is_loggedin(request)
+
+    if user_is_loggedin:
+        return UserResponse(status_code=401, message="User Already Logged in", data=None)
     
+    # if "username" in request.session and "logged_in" in request.session:
+    #     return UserResponse(status_code=401, message="User Already Logged in", data=None)
+
```
Comment (zxenonx) [inline app/routes/auth_routes.py:121]: Kindly Remove commented out code

File: app/routes/auth_routes.py [hunk 6]
```diff
@@ -69,12 +133,103 @@ async def login_user(user: UserRequest, db: Session= Depends(get_db)):
...
+    # Example: Use 'httpx' library to make a GET request to Facebook API
+    # Parse the response and return user data
+    return {"id": 1, "username": "facebook_user"}
+
+# Function to fetch user info from Google using the token
+async def get_google_user(token: str):
+    # Implement the logic to retrieve user data from Google API using 'token'
+    # Example: Use 'httpx' library to make a GET request to Google API
+    # Parse the response and return user data
+    return {"id": 2, "username": "google_user"}
```
Comment (zxenonx) [inline app/routes/auth_routes.py:177]: These ones are not yet ready right?

File: app/routes/auth_routes.py [hunk 7]
```diff
@@ -69,12 +133,103 @@ async def login_user(user: UserRequest, db: Session= Depends(get_db)):
...
+        return {"message": "Facebook Login Successful", "user": user}
     else:
-        return UserResponse(status_code=200, message="Login Successful", data=None)
\ No newline at end of file
+        raise HTTPException(status_code=401, detail="Facebook login failed")
+
+# Google Login Route
+@auth_router.get('/google_login/')
+async def login(request: Request):
+    # print("Hello")
```
Comment (zxenonx) [inline app/routes/auth_routes.py:200]: Kindly remove this

File: app/routes/auth_routes.py [hunk 8]
```diff
@@ -69,12 +133,103 @@ async def login_user(user: UserRequest, db: Session= Depends(get_db)):
...
+#         if not existing_user:
+#             # Create a new user if not found
+#             new_user = User(username=user["username"])
+#             db.add(new_user)
+#             db.commit()
+#             db.refresh(new_user)
+#         db.close()
+#         return {"message": "Google Login Successful", "user": user}
+#     else:
+#         raise HTTPException(status_code=401, detail="Google login failed")
```
Comment (zxenonx) [inline app/routes/auth_routes.py:235]: Should be removed

**Review by zxenonx — CHANGES_REQUESTED** [review]
Thanks @adelajaOlamilekan 
I left a few comments.

File: app/routes/auth_routes.py (hunk 6 above)
Comment (adelajaOlamilekan) [reply app/routes/auth_routes.py:177]: yes they are not

**Review by zxenonx — APPROVED** [review]
Good.

**Pull request merged**